### PR TITLE
Helpful changes in MemoryMap for use in CHAI

### PR DESCRIPTION
**Pull request by jdahm**
Makes `MemoryMap` methods more like `std::map`, and  adds `std::lock_guard<T>`.

- [X] `remove` -> `erase`
- [x] Make threadsafe
- [x] Run all tests
- [x] Check performance against develop
- [X] Add a CHANGELOG.md entry

### CODE REVIEW

File: src/umpire/util/MemoryMap.inl [hunk 1]
```diff
@@ -59,7 +59,7 @@ MemoryMap<V>::~MemoryMap()
 template <typename V>
 template <typename... Args>
 std::pair<typename MemoryMap<V>::Iterator, bool>
-MemoryMap<V>::get(void* ptr, Args&&... args) noexcept
+MemoryMap<V>::get(Key ptr, Args&&... args) noexcept
```
Comment (davidbeckingsale) [inline src/umpire/util/MemoryMap.inl]: Was the `void*` just laziness before? Or does this introduce any real changes?

Comment (jdahm) [reply src/umpire/util/MemoryMap.inl]: It's the same thing. `using Key = void*;`

**Comment by jdahm** [comment]
I checked and the performance is equivalent to that on `develop`

File: src/umpire/util/AllocationMap.cpp [hunk 2]
```diff
@@ -265,20 +268,28 @@ bool AllocationMap::contains(void* ptr) const
...
+AllocationMap::size() const
+{
+  return m_size;
+}
 
 void
 AllocationMap::print(const std::function<bool (const AllocationRecord&)>&& pred,
                      std::ostream& os) const
 {
+  // Aquire lock
```
Comment (davidbeckingsale) [inline src/umpire/util/AllocationMap.cpp]: This is pedantic, but I think these comments above every `lock_guard` are redundant.

Comment (jdahm) [reply src/umpire/util/AllocationMap.cpp]: ok! I'll remove them

Comment (davidbeckingsale) [reply src/umpire/util/AllocationMap.cpp]: thanks :)

**Comment by jdahm** [comment]
This is ready to be reviewed!

**Pull request merged**